pkgs/applications/version-management/sapling/default.nix

```diff
@@ -0,0 +1,166 @@
+    sha256 = "sha256-lRvZpkcxxET9kHpc4mhUMCAIaml/a+CPfMLJp1KieMU=";
+  };
+  Cython = fetchurl {
+    url = "https://files.pythonhosted.org/packages/bc/fa/8604d92ef753e0036d807f1b3179813ab2fa283e3b19c926e11673c8205b/Cython-0.29.26.tar.gz";
+    sha256 = "sha256-rzd9VDp2KGfaEfz25Vj3pKU1/4aT8wzOEj+rEMAPoxI=";
+  };
+in
+python3Packages.buildPythonPackage rec {
+  pname = "sapling";
+  version = "0.1.20221118-210929-cfbb68aa";
+
+  src = fetchFromGitHub {
+    owner = "facebook";
+    repo = pname;
+    rev = version;
+    hash = "sha256-IzbUaFrsSMojhsbpnRj1XLkhO9V2zYdmmZls4mtZquw=";
+  };
+
+  sourceRoot = "source/eden/scm";
+
+  # Upstream does not commit Cargo.lock, which buildRustPackage needs.
+  # postPatch is needed because lockFileContents doesn't copy the lockfile
+  # to src.
+  cargoDeps = rustPlatform.importCargoLock {
+    lockFile = ./Cargo.lock;
+    outputHashes = {
+      "cloned-0.1.0" = "sha256-c3CPWVjOk+VKBLD6WuaYZvBoKi5PwgXmiwxKoCk0bsI=";
+      "deltae-0.3.0" = "sha256-a9Skaqs+tVTw8x83jga+INBr+TdaMmo35Bf2wbfR6zs=";
+      "fb303_core-0.0.0" = "sha256-yoKKSBwqufFayLef2rRpX5oV1j8fL/kRkXBXIC++d7Q=";
+      "fbthrift-0.0.1+unstable" = "sha256-jtsDE5U/OavDUXRAE1N8/nujSPrWltImsFLzHaxfeM0=";
+      "reqwest-0.11.11" = "sha256-uhc8XhkGW22XDNo0qreWdXeFF2cslOOZHfTRQ30IBcE=";
+      "serde_bser-0.3.1" = "sha256-KCAC+rbczroZn/oKYTVpAPJl40yMrszt/PGol+JStDU=";
+    };
+  };
+  postPatch = ''
+    cp ${./Cargo.lock} Cargo.lock
+  '';
+
+  # TODO: Automate Sapling build dep fetching
+  # Since the derivation builder doesn't have network access to remain pure,
+  # fetch the artifacts manually and link them. Then replace the hardcoded URLs
+  # with filesystem paths for the curl calls
+  postUnpack = ''
+    mkdir $sourceRoot/fetched_build_deps
+    ( cd $sourceRoot/fetched_build_deps
+      ln -s ${ptyprocess} ptyprocess-0.7.0-py2.py3-none-any.whl
+      ln -s ${pexpect} pexpect-4.8.0-py2.py3-none-any.whl
+      ln -s ${ipython} ipython-7.16.1-py3-none-any.whl
+      ln -s ${simplegeneric} simplegeneric-0.8.1.zip
+      ln -s ${toml} toml-0.10.2-py2.py3-none-any.whl
+      ln -s ${colorama} colorama-0.4.4-py2.py3-none-any.whl
+      ln -s ${backcall} backcall-0.2.0-py2.py3-none-any.whl
+      ln -s ${setuptools} setuptools-57.0.0-py3-none-any.whl
+      ln -s ${wcwidth} wcwidth-0.2.5-py2.py3-none-any.whl
+      ln -s ${prompt_toolkit} prompt_toolkit-2.0.10-py3-none-any.whl
+      ln -s ${decorator} decorator-5.0.9-py3-none-any.whl
+      ln -s ${pickleshare} pickleshare-0.7.5-py2.py3-none-any.whl
+      ln -s ${Pygments} Pygments-2.9.0-py3-none-any.whl
+      ln -s ${traitlets} traitlets-4.3.3-py2.py3-none-any.whl
+      ln -s ${six} six-1.16.0-py2.py3-none-any.whl
+      ln -s ${ipython_genutils} ipython_genutils-0.2.0-py2.py3-none-any.whl
+      ln -s ${ipdb} ipdb-0.13.9.tar.gz
+      ln -s ${Cython} Cython-0.29.26.tar.gz )
```

Comment:
How hard would it be to rip this all out and use our pythonPackages?

In the current state this is unmaintainable and we can't merge it like this.

Reply:
Agreed, wanted to automate this part with a script; it's hard to tell what they're doing with these deps, but the `setup.py` logic depends on the actual archive instead of the source within. I mostly mimicked the example for `tokenizer` in doing this, to get it working. Are there any alternatives to generating these (and the associated items in the `let` block`)?

Reply:
My current plan is to generate a `deps.nix` file using an update script that can be imported by this file, to abstract away all the crud currently here. Is there a way to calculate the sha256 of a derivation that would be generated from `fetchurl` beforehand, rather than attempting with a fake hash and then correcting it?

Reply:
Ok, I've done the above and made a usable script for automating asset fetching, plus discovered `nix-prefetch-url`. Will integrate the latter into said script and push tomorrow or so, as it's gotten late here.

Reply:
Why do we need the exact versions at all? Can't we just use any version?

Reply:
It probably depends; some versions could be compatible, some could be incompatible. Presumably the versions are pinned because of [version constraints on what has been vendored in FB infra][1]. I can't speak to the Python stuff as much, but an example is in the Rust deps; they use `clap` v2, whereas that package's upstream is on v3 which is breaking, and the [reason was such (internal version constraints)][2].

The weird thing is that all of these assets are downloaded imperatively in archive form, and then processed by the setup script. Usings Nixpkgs versions of these packages IMO would be quite the uphill battle, as (I think?) they'd need to be rearchived into a form the `setup.py` script is expecting, and more substantial change to `setup.py` would need to also happen to rip out the `curl` calls and processing. This wouldn't buy us much over generating `deps.nix`, and also have the cost of possibly introducing bugs related to dependency mismatch. With this PR's method, the binary should be the same as the one created by upstream.

[1]: https://github.com/facebook/sapling/blob/main/eden/scm/setup.py#L410-L424
[2]: https://news.ycombinator.com/item?id=33619406

Reply:
Using the Cargo.lock file for rust is fine but for python we usually take the version constraints very lightweight. 

I don't know who thought it would be a good idea to replace pip with a custom download mechanism. Maybe we should ask upstream how they imagine distros should handle that.

Reply:
This is also a bit worrying to me. Part of the problem is that we're going to need *continuous* integration and testing with upstream's builds in order to make sure this works, and keeps working. It's one of those cases where if you update regularly you can catch small things, but imagine updating this 6 months down the line; it could very well evolve into a nightmare. But either case isn't ideal: it basically means we have to continuously track development, and that's a serious burden, especially when Sapling probably has, I'd assume, half a dozen engineers working on it actively (so somewhere between ~6-times and ~infinity-times more horsepower/manpower when compared to our package, in practice. I'm happy to help co-maintain this expression but I do have limits on how much pain I can tolerate.)

This leads into another issue: quality control. It is *very* possible we have gotten things wrong here that won't crop up until further down the line. There's a very real and big problem that can happen when downstream distros try to deviate too much from what upstream intends to happen. You literally want your version control system to be the *last* thing that does this, because it's very possible to end up in dangerous or extreme situations where e.g. you will desperately need a fix before you can continue your work at all.

I'm not saying that we should shelve this PR, or that it doesn't work as it is. But I think caution and communication with upstream is needed here.

---

@pbar1 Concurrently while we keep working on this, do you think it would be possible to use your intra-Meta connections to ask one of the developers to pop by here? I get the impression you aren't on the Sapling team, and having someone who has an idea of where the project is headed at the moment would be great.

I ask since I remember reading the HN thread and one of the developers stopped by saying things like they are continuing to target more parts of Sapling in Rust than Python for various reasons. This kind of stuff is what I want to know, it's *extremely* relevant to us if Python will keep shedding its skin and get replaced. That would mean some complexity now is more tolerable (because over time we'll get to remove it.)f

There are other things. I strongly suspect Nixpkgs is the first system trying to package Sapling externally, due to its complex dependency chain and Nix's ability to handle that.[1] I also suspect Sapling was overwhelmingly designed for Meta's context (not really a suspicion, to be fair) and in practice *many* software developers aren't aware of the constraints package maintainers are under &mdash (no shade to the Sapling developers, it's something most don't need to be aware of.) I think it would be very good for us to communicate with them if only to set some expectations of what we think would be good, versus what the plans of the Sapling team are.

If we could get some basic communication going about expectations on both sides &mdash; I think that would be really valuable here for both sides.

In the mean time I'll try to address any minor nits I see and test this expression, but I would love to actually get any input from the Sapling team here, see what our priorities are, etc.

> [1] Worth mentioning most systems would give up on the hopeless complexity here but Nix can actually handle it, so we should count our blessings but remember, nothing is free.

Reply:
And just to be abundantly clear: I am very, *very* eager to have Sapling in Nixpkgs and on all my Nix-based machines. Itr's truly wonderful (and feel free to share that with your colleagues.) So don't take my caution this as a NACK on this PR or anything like that. I just want to make sure we aren't going to tie ourselves into a knot six months down the line with mismatched priorities.

Reply:
I suspect there is a needless amount of debate which distracts from the end goal. I know Nix is a build paltform and everyone feels good when they know builds better than the other guy but end of the day a project needs working binaries, and sapling binary works. 

Perhaps, just perhaps, building in the most ideal way isn't the priority for fb team right now? Maybe 6 months down the line they'd be using the latest and best, maybe ship a default.nix and flake.nix? 

I know I sound very bitter but I see this attitude repeatedly where PR creators are punished for their attempts. OP has a working binary, great, lets go with it. I am all for maintainability but hey, there are no spaces at the end of newline so that should be good enough!

Reply:
> I suspect

Your suspicions are wrong.

> I know

No, you clearly don't.

> Perhaps, just perhaps, building in the most ideal way isn't the priority for fb team right now?

Where did someone say it should *be* their priority? I'm getting pretty sick of people who do not contribute on this repository beyond half-assed bug reports, if that, constantly shoving words in the mouths of developers, all because they're in a rush to complain we aren't doing free work for them *fast enough*. Sapling was released *a week ago*. We're lucky someone even stepped up to try this early given my initial glance at the complexity of the package. And you still complaining that we still haven't merged this fast enough, still?

My caution is based on the fact this is a complex expression with generated dependencies, *and those kinds of expressions have high costs to maintain, and are potentially risky for users* because they introduce users to risk that the original developers may not anticipate, if not done very carefully. This is a version control system. Risk factors are potentially extreme. And if that risk factor blows up, users both complain to the Sapling developers &mdash; who have no fucking clue what to expect because downstream introduced unexpected changes &mdash; and *also* to Nix developers. So we then have to write a shitload of code to handle that in terms of things like dependency generators, careful patching, etc. We literally have to write our own `Cargo.lock` file because the internal Meta infrastructure uses vendored deps we can't see! This is a *huge* source of potential issues to anyone who has experience in this area. It's largely me being cautious, *not* demanding, and again, it's based on experience. Try running `git log --author="Austin Seipp" --format=oneline --no-merges | wc -l` and you'll have an idea of how much experience that is.

The Sapling team is not obligated to make changes we need; in fact if they said "No" to everything we wanted here, *that would still be informative for our purposes so we know what to expect*. This is good! This is literally completely self-evident to anyone who has had to work between teams before! Literally this happens *all the time* in open source. Another good example of a relevant question is "Are you going to do a stable release anytime soon or should we expect to use unstable tags for the forseeable future?" And given that Eden and Mononoke have yet to be released, these concerns are likely going to *increase* for us and the Sapling team. These are relevant questions to maintainers who are responsible for delivering the software to Nix users.

Again: The current expression works. It basically *looks fine*, too, assuming we're on board with all this. I never said I wasn't. I think we *should* have communication with them to make it easy as possible to maintain this in the future, especially wrt concerns like `Cargo.lock`. In my experience communicating with upstreams, I think this will in turn give the Sapling developers a little insight into what we and others will need. None of this is in conflict.

> I know I sound very bitter but I see this attitude repeatedly where PR creators are punished for their attempts.

Let me get this straight: you aren't writing the patches, you aren't maintaining the package, you aren't writing the software. The only thing you seem to be doing is posting in a text box about how we aren't doing this to your personal liking and at the speed which personally makes you happy &mdash; yet *you're* the one who is bitter and whose feelings need to be assuaged? Sorry, but that sounds like a You problem, not a Me problem. And my suggested solutions for that problem of yours would be to get a grip, and take it to a therapist, in that order.

Reply:
I've definitely contributed and stopped only because of this attitude

Reply:
Good, we don't need you. And don't let the door hit your ass on the way out.

Reply:
> Part of the problem is that we're going to need continuous integration and testing with upstream's builds in order to make sure this works,

Yes, like for *any* other program in nixpkgs which is not using upstreams vendored dependencies (which no program should) and where upstream does not have really tight version ranges which has its own problems.

> I know Nix is a build paltform

yes and nixpkgs is a package collection that is comparable to the ones used by apt and rpm and using such hacks would be completely unacceptable for the main repository. 

If you need something that "just works" then upstreaming to nixpkgs might not be the option you are searching for for packages like sapling that have such a horrible build process. 

> I know I sound very bitter but I see this attitude repeatedly where PR creators are punished for their attempts. OP has a working binary, great, lets go with it. I am all for maintainability but hey, there are no spaces at the end of newline so that should be good enough!

nixpkgs has a higher standard then things working. Otherwise managing security patches would be impossible. If this is the best solution we have right now then I'd rather not have sapling.

I just skimmed through @thoughtpolice posts but this PR should have not be merged in this state and I am voting for reverting it rather than having this compromise.